lib/commands/exam.js

```diff
@@ -254,5 +372,97 @@ module.exports = TestCommand.extend({
+        const browserId = /browser=\s*([0-9]*)/.exec(this.launcher.settings.test_page)[1];
+
+        if (browserId === undefined) {
+          throw new Error('Browser Id can\'t be found.');
+        }
+
+        this.moduleQueue = browserModuleMapping[browserId];
+      } else if (!proto.moduleQueue) {
+        proto.moduleQueue = modules;
+        proto.moduleMap = {};
+      }
+    }
+
+    events['get-next-module'] = function() {
+      const proto = Object.getPrototypeOf(this);
+      const replayExecution = this.config.progOptions.browser_module_mapping;
+      const moduleQueue = this.moduleQueue;
+      const moduleName = moduleQueue.shift();
+
+      if (!moduleName) {
+        this.socket.emit('testem:module-queue-complete');
+      } else {
+        this.socket.emit('testem:next-module-response', moduleName);
+        // keep track of the modules ran per browserId when appMode is ci and replayExecution flag is false
+        if (this.config.appMode === 'ci' && !replayExecution) {
+          const browserId = /browser=\s*([0-9]*)/.exec(this.launcher.settings.test_page)[1];
+
+          if (browserId === undefined) {
+            throw new Error('Browser Id can\'t be found.');
+          }
+
+          proto.moduleMap[browserId] = proto.moduleMap[browserId] || [];
+          proto.moduleMap[browserId].push(moduleName);
+        }
+      }
+
+    }
+
+    if (this.validator.shouldReplayExecution) {
+      return events;
+    }
+
+    // We only need to register `after-tests-complete` event to create json file when the test execution run is not for replay-execution.
+    events['after-tests-complete'] = function() {
+      const proto = Object.getPrototypeOf(this);
+      const browserCount = Object.keys(config.testPage).length;
+      // custom_browser_socket_events is set in config when _loadBalance is in test pages and browser_module_mapping is set when running `replay-execution`.
+      const shouldLoadBalance = !!this.config.progOptions.custom_browser_socket_events && !this.config.progOptions.browser_module_mapping
+
+      proto.finishedBrowsers = proto.finishedBrowsers ? proto.finishedBrowsers + 1 : 1;
+
+      if (proto.finishedBrowsers == browserCount) {
+        // Clean the moduleQueue and finishedBrowser. This guarantees new test run can set the moduleQueue. Otherwise, the server will return no more test to run.
+        // TODO: Investigate not using proto
+        delete proto.moduleQueue;
+        delete proto.finishedBrowser;
+        if (this.config.appMode === 'ci' && shouldLoadBalance) {
+          const fileName = `test-execution-${moment().format('YYYY-MM-DD__HH-MM-SS')}.json`;
+          const moduleMapJson = {
+            numberOfBrowsers: browserCount,
+            executionMapping: proto.moduleMap
+          };
+          const testExecutionJson = JSON.stringify(moduleMapJson);
+          fs.writeFileSync(fileName, testExecutionJson, (err) => {
```

Comment:
`fs.writeFileSync` does not have a 3rd callback for error handling: https://nodejs.org/api/fs.html#fs_fs_writefilesync_file_data_options